Doc for Konnectivity service feature (#19216)
kubernetes/website: content/en/docs/concepts/architecture/master-node-communication.md

know what you are doing. The Konnectivity service is a replacement for this
communication channel.

### Konnectivity service
{{< feature-state for_k8s_version="v1.18" state="beta" >}}

As a replacement to the SSH tunnels, the Konnectivity service provides TCP
level proxy for the Master → Cluster communication. The feature is beta
in 1.18.

To use it on GCE, you can deploy the cluster using kube-up.sh in one line:

```bash
KUBE_ENABLE_EGRESS_VIA_KONNECTIVITY_SERVICE=true ./cluster/kube-up.sh
```

Let's look at what kube-up.sh does, so that you can use the Konnectivity service
in other environments with some modifications.

First, you need to configure the API Server to use the Konnectivity service
to direct its network traffic to cluster nodes:
1. Set the `--egress-selector-config-file` flag of the API Server, it is the
path to the API Server egress configuration file.
2. At the path, create a configuration file. For example,

```yaml
apiVersion: apiserver.k8s.io/v1beta1
kind: EgressSelectorConfiguration
egressSelections:
# Since we want to control the egress traffic to the cluster, we use the
# "cluster" as the name. Other supported values are "etcd", and "master".

> - name: cluster

Reviewer: We should include master and etcd and just note that they should be kept at their default values of direct connection. They "technically" need to be defined.

Author: I omitted them because this doc is about master->cluster traffic. And omitting "etcd" and "master" is ok, because by default they would do direct dial. e.g., [etcd dial](https://github.com/kubernetes/kubernetes/pull/82048/files#diff-d02b855276b48d62bd9ec183560db2e2R126).

  connection:
    # This controls the protocol between the API Server and the Konnectivity
    # server. Supported values are "GRPC" and "HTTPConnect". There is no
    # end user visible difference between the two modes. You need to set the
    # Konnectivity server to work in the same mode.
    proxyProtocol: GRPC
    transport:
      # This controls what transport the API Server uses to communicate with the
      # Konnectivity server. UDS is recommended if the Konnectivity server
      # locates on the same machine as the API Server. You need to configure the
      # Konnectivity server to listen on the same UDS socket.
      # The other supported transport is "tcp". You will need to set up TLS 
      # config to secure the TCP transport.
      uds:
        udsName: /etc/srv/kubernetes/konnectivity-server/konnectivity-server.socket

> ```

Reviewer 2: If this is a valid file that a reader could apply directly, make it downloadable. See https://kubernetes.io/docs/contribute/style/write-new-topic/#including-code-from-another-file

Next, you need to deploy the Konnectivity service server and agents.
[kubernetes-sigs/apiserver-network-proxy](https://github.com/kubernetes-sigs/apiserver-network-proxy)
is a reference implementation.

> Deploy the Konnectivity server on you master node:

Reviewer 2: nit: “you” should be “your”.

You could reword:
```suggestion
Deploy the Konnectivity server into your cluster's control plane. If your control plane runs Kubernetes components inside Pods, you can run a DaemonSet that selects the control plane nodes:
```
(and then skip to defining a DaemonSet rather than a Pod).

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: konnectivity-server
  namespace: kube-system
spec:
  priorityClassName: system-cluster-critical
  hostNetwork: true
  containers:
  - name: konnectivity-server-container
    image: us.gcr.io/k8s-artifacts-prod/kas-network-proxy/proxy-server:v0.0.8
    command: ["/proxy-server"]
    args: [
            "--log-file=/var/log/konnectivity-server.log",
            "--logtostderr=false",
            "--log-file-max-size=0",
            # This needs to be consistent with the value set in egressSelectorConfiguration.
            "--uds-name=/etc/srv/kubernetes/konnectivity-server/konnectivity-server.socket",

> 	    # The following two lines assume the Konnectivity server is

Reviewer: nit: indentation

            ]
    livenessProbe:
      httpGet:
        scheme: HTTP
        host: 127.0.0.1
        port: 8133
        path: /healthz
      initialDelaySeconds: 30
      timeoutSeconds: 60
    ports:
    - name: agentport
      containerPort: 8132
      hostPort: 8132
    - name: adminport
      containerPort: 8133
      hostPort: 8133
    volumeMounts:
    - name: varlogkonnectivityserver
      mountPath: /var/log/konnectivity-server.log
      readOnly: false
    - name: pki
      mountPath: /etc/srv/kubernetes/pki
      readOnly: true
    - name: konnectivity-uds
      mountPath: /etc/srv/kubernetes/konnectivity-server
      readOnly: false
  volumes:
  - name: varlogkonnectivityserver
    hostPath:
      path: /var/log/konnectivity-server.log
      type: FileOrCreate
  - name: pki
    hostPath:
      path: /etc/srv/kubernetes/pki
  - name: konnectivity-uds
    hostPath:
      path: /etc/srv/kubernetes/konnectivity-server
      type: DirectoryOrCreate
```

Then deploy the Konnectivity agents in your cluster:

```yaml
apiVersion: apps/v1
# Alternatively, you can deploy the agents as Deployments. It is not necessary

> # to have an agnet on each node.

Reviewer: s/agnet/agent

kind: DaemonSet
metadata:
  labels:
    addonmanager.kubernetes.io/mode: Reconcile
    k8s-app: konnectivity-agent
  namespace: kube-system
  name: konnectivity-agent
spec:
  selector:
    matchLabels:
      k8s-app: konnectivity-agent
  template:
    metadata:
      labels:
        k8s-app: konnectivity-agent
    spec:
      priorityClassName: system-cluster-critical
      tolerations:
        - key: "CriticalAddonsOnly"
          operator: "Exists"
      containers:
        - image: us.gcr.io/k8s-artifacts-prod/kas-network-proxy/proxy-agent:v0.0.8
          name: konnectivity-agent
          command: ["/proxy-agent"]
          args: [
                  "--logtostderr=true",
                  "--ca-cert=/var/run/secrets/kubernetes.io/serviceaccount/ca.crt",

>                   "--proxy-server-host=35.225.206.7",

Reviewer: nit: Maybe include a note to indicate how to obtain this address?

                  "--proxy-server-port=8132",
                  "--service-account-token-path=/var/run/secrets/tokens/konnectivity-agent-token"
                  ]
          volumeMounts:
            - mountPath: /var/run/secrets/tokens
              name: konnectivity-agent-token
          livenessProbe:
            httpGet:
              port: 8093
              path: /healthz
            initialDelaySeconds: 15
            timeoutSeconds: 15
      serviceAccountName: konnectivity-agent
      volumes:
        - name: konnectivity-agent-token
          projected:
            sources:
              - serviceAccountToken:
                  path: konnectivity-agent-token
                  audience: system:konnectivity-server
```

> Last, create the relevant RBAC rules:

Reviewer 2: Does this assume that the cluster is using RBAC? As I understand it RBAC is recommended but still optional: see https://kubernetes.io/docs/reference/access-authn-authz/rbac/#upgrading-from-abac

Author: Updated to "Last, if RBAC is enabled in your cluster, create the relevant RBAC rules"

### SSH Tunnels

Kubernetes supports SSH tunnels to protect the Master → Cluster communication
paths. In this configuration, the apiserver initiates an SSH tunnel to each node
in the cluster (connecting to the ssh server listening on port 22) and passes
all traffic destined for a kubelet, node, pod, or service through the tunnel.
This tunnel ensures that the traffic is not exposed outside of the network in
which the nodes are running.

SSH tunnels are currently deprecated so you shouldn't opt to use them unless you
know what you are doing. The Konnectivity service is a replacement for this
communication channel.

### Konnectivity service
{{< feature-state for_k8s_version="v1.18" state="beta" >}}

As a replacement to the SSH tunnels, the Konnectivity service provides TCP
level proxy for the Master → Cluster communication. The Konnectivity consists of
two parts, the Konnectivity server and the Konnectivity agents, running in the
Master network and the Cluster network respectively. The Konnectivity agents
initiate connections to the Konnectivity server and maintain the connections.
All Master → Cluster traffic then goes through these connections.

> See [Konnectivity Service Setup](/docs/tasks/setup-konnectivity/) on how to set

Reviewer 3: @caesarxuchao , Is there a task page?

Author: The task page is in https://github.com/kubernetes/website/pull/19760, per https://github.com/kubernetes/website/pull/19216#discussion_r395861817 by @sftim.

Reviewer 2: > move the the link to #19760 as well

That sounds like a good idea.

This tunnel ensures that the traffic is not exposed outside of the network in
which the nodes are running.

SSH tunnels are currently deprecated so you shouldn't opt to use them unless you know what you are doing. APIServer network proxy is a replacement for this communication channel.

> ### APIServer network proxy (alpha)

Reviewer 2: I recommend against putting feature state in links as they then aren't stable
```suggestion
### API Server network proxy
```
and add a [feature state shortcode](https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#feature-state) for the section.

```suggestion
### API Server network proxy (alpha) {#apiserver-network-proxy}
```
also works, but I'm less keen.

Reviewer 4: Didn't this make beta?

As a replacement to the SSH tunnels, Kubernetes supports APIServer network proxy

> to protect the Master -> Cluster communication paths. The feature is still

Reviewer 2: ```suggestion
to protect the control plane → cluster communication. The feature is still
```
If you use a feature state shortcode, you don't need to repeat that detail in the text.

Author: What do you mean by "feature state shortcode"?

Reviewer 2: Ah, sorry. The heading `### APIServer network proxy (alpha)` repeats the state from the [feature state shortcode](https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#feature-state).

There's another problem: headings generate anchor URLs, and we plan to graduate this feature to GA in the future. So we should avoid making a URL that has “alpha” in it.

alpha. It supports deploying on GCE via the

> [kube-up.sh](https://github.com/kubernetes/kubernetes/blob/master/cluster/kube-up.sh)

Reviewer 4: This section should go with the documentation for kube-up.sh, no?

script. Run `KUBE_ENABLE_EGRESS_VIA_KONNECTIVITY_SERVICE=true
./cluster/kube-up.sh` to start the cluster with APIServer network proxy on.
You can find more details in the [apiserver-network-proxy

> repository](https://github.com/kubernetes-sigs/apiserver-network-proxy).

Reviewer 2: This seems very specific to a particular cloud infrastructure provider. Instead, link to https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/ (search for links to `/docs/reference/command-line-tools-reference/feature-gates/` to see how other pages typically do this); readers can then work out what approach they need to enable the feature manually.

Author: Unfortunately this code is not behind a feature flag. It's is enabled/disabled by this environment variable in kube-up.sh.

Reviewer 4: I think you need to explain how someone might set this up and use it, e.g. explain the apiserver settings, the control plane proxy binary, and the agent.
